teps/0156-whenexpressions-in-step.md

```diff
@@ -0,0 +1,235 @@
+---
+status: proposed
+title: WhenExpressions in Steps
+creation-date: '2024-04-15'
+last-updated: '2024-04-15'
+authors:
+- '@ericzzzzzzz'
+---
+
+# TEP-0156: WhenExpressions in Steps
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+   - [Goals](#goals)
+   - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+   - [When Expressions Evaluation](#when-expressions-evaluation)
+   - [Supported Expression Languages](#supported-expression-languages)
+- [Test Plan](#test-plan)
+- [Alternatives](#alternatives)
+
+## Summary
+This TEP proposes the addition of When Expressions to individual Steps within a Task. This will allow for fine-grained control over the execution of steps based on conditions like the status of previous step results. This enhancement will provide greater flexibility and control in defining task logic, enabling more complex and dynamic workflows.
+## Motivation
+Currently, Tekton Tasks execute steps sequentially without the ability to control their execution based on conditions within the Task itself. This limitation restricts the ability to create dynamic workflows that adapt to different scenarios or respond to the outcomes of previous steps.
+By introducing When Expressions to Steps, users will be able to:
+ - **Create conditional workflows**: Execute steps only when specific conditions are met, such as the success or failure of a previous step, the value of a result, or the presence of a specific parameter.
+ - **Improve error handling**: Implement robust error handling strategies by executing specific steps only when errors occur in previous steps.
+ - **Enhance reusability**: Develop more versatile Tasks that can adapt their behavior based on different input parameters or runtime conditions.
+
+## Goals
+ - Enable the use of When Expressions to control the execution of individual Steps within a Task.
+## Non-Goals
+ - Introduce new expression syntax or operators beyond what is already available in When Expressions.
+
+## Proposal
+We propose extending the Step definition in Tekton Tasks to include an optional when field, similar to the existing implementation for Tasks and Pipelines. The when field will accept a list of When Expressions that define the conditions for executing the Step.
+
+### When Expressions Evaluation
+
+When expressions are evaluated at the entrypointer level. This allows us to use step results produced from previous steps to determine if a step should be skipped. To indicate a skipped step, the exit code will be 0 (showing a step is complete) and the TerminationReason will be Skipped.
```

Comment:
Do we forsee any impact on the entrypoint "performance" ? The entrypoint is doing more and more work (with step results, …) and I think we never measured or "thought" about possible performance impact (and if they are noticeable or not) ?

Reply:
You're right, adding Step-level When Expressions will have some cost, but it's important to consider the value it brings. This cost is only incurred when Step-level When Expressions are actually configured. In comparison to the benefits of having this finer-grained control over workflow and the potential to skip long-running steps for performance optimization, the cost is relatively minimal. Even Task-level When Expressions also have a performance cost, and the Step-level implementation offers a significantly greater potential for performance gains through more flexible workflow control.